File: fliptheswitch-gradle/src/main/java/com/github/michaelengland/fliptheswitch/gradle/FeaturesExtension.java
package com.github.michaelengland.fliptheswitch.gradle;

import org.gradle.api.Action;
import org.gradle.api.NamedDomainObjectContainer;
import org.gradle.api.Project;

public class FeaturesExtension {
    private NamedDomainObjectContainer<FeatureDefinition> defaultConfig;
    private NamedDomainObjectContainer<ProductFlavorConfig> productFlavors;

    public NamedDomainObjectContainer<FeatureDefinition> defaultConfig() {
        return defaultConfig;
    }

    public void initialize(final Project project) {
        defaultConfig = project.container(FeatureDefinition.class);
        productFlavors = project.container(ProductFlavorConfig.class);
    }

    public void defaultConfig(final Action<NamedDomainObjectContainer<FeatureDefinition>> action) {
        action.execute(defaultConfig);
    }

    public NamedDomainObjectContainer<ProductFlavorConfig> productFlavors() {
        return productFlavors;
    }

    public void productFlavors(final Action<NamedDomainObjectContainer<ProductFlavorConfig>> action) {
        action.execute(productFlavors);
    }
}

File: fliptheswitch-gradle/src/main/kotlin/com/github/michaelengland/fliptheswitch/gradle/FeaturesExtension.kt
package com.github.michaelengland.fliptheswitch.gradle

import org.gradle.api.Action
import org.gradle.api.NamedDomainObjectContainer
import org.gradle.api.Project

open class FeaturesExtension {
    lateinit var defaultConfig: NamedDomainObjectContainer<FeatureDefinition>
    lateinit var productFlavors: NamedDomainObjectContainer<ProductFlavorConfig>

    fun initialize(project: Project) {
        defaultConfig = project.container(FeatureDefinition::class.java)
        productFlavors = project.container(ProductFlavorConfig::class.java)
    }

    fun defaultConfig(action: Action<NamedDomainObjectContainer<FeatureDefinition>>) {
        action.execute(defaultConfig)
    }

    fun productFlavors(action: Action<NamedDomainObjectContainer<ProductFlavorConfig>>) {
        action.execute(productFlavors)
    }
}
